proxy-proto.c

```diff
@@ -0,0 +1,332 @@
+#include "gmid.h"
+
+#include <ctype.h>
+#include <string.h>
+
+#define MIN(a, b) (a) < (b) ? (a) : (b)
+
+static const char* 
+consume_token(
+    const char *buf, 
+    size_t buflen, 
+    const char *token, 
+    size_t tokenlen, 
+    size_t *consumed_len
+) {
+    // buflen may be smaller than tokenlen
+    // in that case, compare until end of buf
+    size_t checklen = MIN(buflen, tokenlen);
+    if (NULL != consumed_len)
+        *consumed_len = checklen;
+
+    return memmem(buf, buflen, token, checklen);
+}
+
+static int 
+check_prefix_v1(const char **buf, size_t *buflen)
+{
+    static const char PROXY[6] = "PROXY ";
+
+    size_t consumed;
+    const char *found = consume_token(*buf, *buflen, PROXY, 6, &consumed);
```

Comment:
What if the first data packet we receive is an incomplete line of the header? Is it still valid to alter the buffer?

Reply:
My thought process was:

1. keep on reading until end of buffer
2. if everything we read is valid parts of the protocol, return `TLS_WANT_POLLIN`
3. if at some point we see something funny, pass the unaltered buffer to libtls
4. if we reach `\r\n`, cut off everything else and pass it to libtls while parsing the rest

Reply:
Ah yes, I forgot to suggest one more thing.  To make parsing easier, we could keep reading until we find CRLF (or fill the buffer, in that case we abort the connection.)

At that point, you have the whole protocol line in memory and parsing becomes easier.  That, together with my other suggestion about changing the underlying string, should make the code easier to follow I believe.

(the rationale is that the proxy protocol is thought to fit one TCP packet so in practice we shouldn't issue more than one read(2), and even if we have it's just 108 bytes at maximum.)

Reply:
Okay, seems reasonable.
Strictly speaking though, the v1 spec says:
> For "UNKNOWN", the rest of the line before the CRLF may be omitted by the sender, and the receiver must ignore anything presented before the CRLF is found.

Which I read as "there does not necessarily need to be a limit to header length for UNKNOWN protocols". But I suppose since there are no "unknown" protocols (yet?) it is fine to ignore that?

Reply:
That's a good point.  The specification says that

> So a 108-byte buffer is always enough to store all the line and a trailing zero for string processing.

but only after showing the maximum line for a non-UNKNOWN protocol, so I can see why there's space for different interpretations.

In any case, I don't think we should go that much out of our way to handle an hypothetical UNKNOWN line with more than 107 bytes of payload, also because I don't really like the idea of keep reading indefinitely for an edge case (I believe) noone is using in practice.

So let's just assume that also UNKNOWN protocols fit the 108 byte buffer.  If we fill the buffer without seeing the CRLF, let's just drop the connection with a nice error in the log and move on.